### PR TITLE
Add new cop `Lint/OrAssignmentToConstant`

### PR DESCRIPTION
**Pull request by uplus**
This cop checks unintended or-assignment to constant.
Usually, you don't need to consider whether A is already assigned elsewhere.

```ruby
# bad
CONST ||= 1

# good
CONST = 1
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] ~Commit message starts with `[Fix #issue-number]` (if the related issue exists).~
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop-hq/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      #   # bad
+      #   CONST ||= 1
+      #
+      #   # good
+      #   CONST = 1
+      #
+      class OrAssignmentToConstant < Base
+        extend AutoCorrector
+
+        MSG = 'Probably unintended or-assignment. Use plain assignment.'
```
Comment (dvandersluis) [inline lib/rubocop/cop/lint/or_assignment_to_constant.rb:20]:
```suggestion
        MSG = 'Avoid use or-assignment with constants.'
```

Might be useful to include the correction in the message as well (<code>Use &#x60;...&#x60; instead.</code>)

File: spec/rubocop/cop/lint/or_assignment_to_constant_spec.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::Lint::OrAssignmentToConstant, :config do
+  subject(:cop) { described_class.new(config) }
+
+  it 'registers an offense when or-assignment to constant.' do
```
Comment (dvandersluis) [inline spec/rubocop/cop/lint/or_assignment_to_constant_spec.rb:6]:
```suggestion
  it 'registers an offense with or-assignment to a constant' do
```

Similarly for the other test descriptions as well.

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks unintended or-assignment to constant.
+      # Usually, you don't need to consider whether A is already assigned elsewhere.
```
Comment (dvandersluis) [inline lib/rubocop/cop/lint/or_assignment_to_constant.rb:7]:
```suggestion
      # This cop checks for unintended or-assignment to a constant.
      # Usually, you don't need to consider whether the constant is already assigned elsewhere.
```

It'd be good to explain why or-assignment for a constant is not a good idea.

Also please add a note about why this cop is not safe.

**Review by dvandersluis — COMMENTED** [review]
Looks like a useful cop!

File: spec/rubocop/cop/lint/or_assignment_to_constant_spec.rb (hunk 2 above)
Comment (uplus) [reply spec/rubocop/cop/lint/or_assignment_to_constant_spec.rb]: Thank you for fixing my English :sweat_smile:  
I fixed all other descriptions.

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb (hunk 3 above)
Comment (uplus) [reply lib/rubocop/cop/lint/or_assignment_to_constant.rb]: OK, I rewrote the description!

**Comment by uplus** [comment]
Thank you for review.
I fixed the three places you pointed out, resolved conflicts, and squashed them.

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb (hunk 1 above)
Comment (uplus) [reply lib/rubocop/cop/lint/or_assignment_to_constant.rb]: I applied the commit-suggestion.

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      #   # bad
+      #   CONST ||= 1
+      #
+      #   # good
+      #   CONST = 1
+      #
+      class OrAssignmentToConstant < Base
+        extend AutoCorrector
+
+        MSG = 'Avoid use or-assignment with constants.'
```
Comment (dvandersluis) [inline lib/rubocop/cop/lint/or_assignment_to_constant.rb:26]:
```suggestion
        MSG = 'Avoid using or-assignment with constants.'
```

Sorry, I made a typo in my previous suggestions! 🤦

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb (hunk 3 above)
Comment (dvandersluis) [reply lib/rubocop/cop/lint/or_assignment_to_constant.rb]: Thanks!

File: lib/rubocop/cop/lint/or_assignment_to_constant.rb (hunk 4 above)
Comment (uplus) [reply lib/rubocop/cop/lint/or_assignment_to_constant.rb]: I applied this commit-suggestion and fixed related spec :+1:

**Review by dvandersluis — APPROVED** [review]
LGTM, thanks :+1:

@rubocop-hq/rubocop-core ci/circleci: cc-upload-coverage keeps failing on this PR, anyone have any ideas why?

**Comment by koic** [comment]
> ci/circleci: cc-upload-coverage keeps failing on this PR, anyone have any ideas why?

It's possible that Code Climate's API wasn't working properly.  I think this failure can be ignored.

```console
#!/bin/bash -eo pipefail
./tmp/cc-test-reporter sum-coverage tmp/codeclimate.*.json --parts 6 --output tmp/codeclimate.total.json
./tmp/cc-test-reporter upload-coverage --input tmp/codeclimate.total.json
Error: response from https://api.codeclimate.com/v1/test_reports.
HTTP 401: You are not authorized for this action
Usage:
  cc-test-reporter upload-coverage [flags]

Flags:
  -s, --batch-size int    batch size for source files (default 500)
  -e, --endpoint string   endpoint to upload coverage information to (default "https://api.codeclimate.com/v1/test_reports")
  -r, --id string         reporter identifier (default "a11b66bfbb1acdf220d5cb317b2e945a986fd85adebe29a76d411ad6d74ec31f")
  -i, --input string      input path (default "coverage/codeclimate.json")
      --insecure          send coverage insecurely (without HTTPS)

Global Flags:
  -d, --debug   run in debug mode


Exited with code exit status 255
CircleCI received exit code 255
```

**Comment by dvandersluis** [comment]
@koic yeah that’s what I figured but I wanted to double check, thanks.

**Pull request merged**